File: java/src/ReactLoop.java
/**
 * Production ReAct Loop — Java
 * Reason → Act → Observe, repeat until finish() called.
 *
 * pom.xml dependencies:
 *   com.squareup.okhttp3:okhttp:4.12.0
 *   com.fasterxml.jackson.core:jackson-databind:2.17.0
 *
 * export ANTHROPIC_API_KEY=sk-ant-...
 * javac -cp okhttp*.jar:jackson*.jar ReactLoop.java
 * java  -cp .:okhttp*.jar:jackson*.jar ReactLoop "your task here"
 */

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.*;
import okhttp3.*;

import java.io.*;
import java.nio.file.*;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.regex.Pattern;

public class ReactLoop {

    // ── Constants ─────────────────────────────────────────────────────────────
    static final String API_URL    = "https://api.anthropic.com/v1/messages";
    static final String MODEL      = "claude-opus-4-6";
    static final ObjectMapper JSON = new ObjectMapper();

    static final OkHttpClient HTTP = new OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build();

    static final String REACT_SYSTEM =
        "You are an autonomous agent operating in a ReAct loop (Reason + Act + Observe).\n\n" +
        "For every task:\n" +
        "1. Use the think tool to reason about what to do next before acting.\n" +
        "2. Call the appropriate tool to act.\n" +
        "3. Observe the result and reason again.\n" +
        "4. Repeat until you have a complete, verified answer.\n" +
        "5. Call finish with your final answer when done.\n\n" +
        "Rules:\n" +
        "- Always think before acting. Never skip the think step.\n" +
        "- If a tool errors, reason about why and try a different approach.\n" +
        "- Do not guess. If unsure, search or fetch.\n" +
        "- Be thorough. Do not call finish until the task is fully complete.";

    // ── Trace types ───────────────────────────────────────────────────────────
    record TraceStep(String type, String content, String tool,
                     JsonNode input, boolean error) {
        static TraceStep thought(String content) {
            return new TraceStep("thought", content, null, null, false);
        }
        static TraceStep action(String tool, JsonNode input) {
            return new TraceStep("action", null, tool, input, false);
        }
        static TraceStep observation(String content, boolean error) {
            return new TraceStep("observation", content, null, null, error);
        }
        static TraceStep answer(String content) {
            return new TraceStep("answer", content, null, null, false);
        }
        ObjectNode toJson() {
            ObjectNode n = JSON.createObjectNode();
            n.put("type", type);
            if (content != null) n.put("content", content);
            if (tool != null)    n.put("tool", tool);
            if (input != null)   n.set("input", input);
            if (error)           n.put("error", true);
            return n;
        }
    }

    static class Trace {
        List<TraceStep> steps       = new ArrayList<>();
        long            totalTokens = 0;
        int             iterations  = 0;
        long            elapsedMs   = 0;
    }

    // ── Tool registry ─────────────────────────────────────────────────────────
    interface ToolFn extends Function<JsonNode, String> {}

    record ToolDef(String name, String description, JsonNode schema, ToolFn fn) {}

    static final Map<String, ToolDef> TOOLS = new LinkedHashMap<>();

    static void registerTool(String name, String description, JsonNode schema, ToolFn fn) {
        TOOLS.put(name, new ToolDef(name, description, schema, fn));
    }

    static ArrayNode toolSchemas() {
        ArrayNode arr = JSON.createArrayNode();
        for (ToolDef t : TOOLS.values()) {
            ObjectNode n = JSON.createObjectNode();
            n.put("name", t.name());
            n.put("description", t.description());
            n.set("input_schema", t.schema());
            arr.add(n);
        }
        return arr;
    }

    // ── Built-in tools ────────────────────────────────────────────────────────
    static {
        // think
        registerTool("think",
            "Reason step by step before acting. Does not call external systems.",
            schema("object", Map.of("reasoning", Map.of("type", "string")), List.of("reasoning")),
            input -> "OK"
        );

        // search
        registerTool("search",
            "Search the web for current information.",
            schema("object", Map.of("query", Map.of("type", "string")), List.of("query")),
            input -> {
                // Replace with Brave, Tavily, SerpAPI, etc.
                throw new RuntimeException("Search not wired up. query: " + input.path("query").asText());
            }
        );

        // fetch_url
        registerTool("fetch_url",
            "Fetch the text content of a URL.",
            schema("object", Map.of("url", Map.of("type", "string")), List.of("url")),
            input -> {
                String url = input.path("url").asText();
                Request req = new Request.Builder().url(url).build();
                try (Response resp = HTTP.newCall(req).execute()) {
                    String body = resp.body() != null ? resp.body().string() : "";
                    // Strip HTML tags
                    String clean = body.replaceAll("<[^>]+>", " ").replaceAll("\\s+", " ").trim();
                    return clean.length() > 8000 ? clean.substring(0, 8000) : clean;
                } catch (IOException e) {
                    throw new RuntimeException("fetch_url failed: " + e.getMessage(), e);
                }
            }
        );

        // run_bash
        registerTool("run_bash",
            "Execute a bash command. Returns stdout/stderr. Timeout 15s.",
            schema("object", Map.of("command", Map.of("type", "string")), List.of("command")),
            input -> {
                String command = input.path("command").asText();
                try {
                    ProcessBuilder pb = new ProcessBuilder("bash", "-c", command);
                    pb.redirectErrorStream(true);
                    Process proc = pb.start();
                    boolean finished = proc.waitFor(15, TimeUnit.SECONDS);
                    if (!finished) {
                        proc.destroyForcibly();
                        return "Command timed out after 15s";
                    }
                    String out = new String(proc.getInputStream().readAllBytes()).trim();
                    if (out.isEmpty()) return "(no output)";
                    return out.length() > 8000 ? out.substring(0, 8000) : out;
                } catch (Exception e) {
                    throw new RuntimeException("run_bash failed: " + e.getMessage(), e);
                }
            }
        );

        // read_file
        registerTool("read_file",
            "Read a file from disk.",
            schema("object", Map.of("path", Map.of("type", "string")), List.of("path")),
            input -> {
                try {
                    return Files.readString(Path.of(input.path("path").asText()));
                } catch (IOException e) {
                    throw new RuntimeException("read_file failed: " + e.getMessage(), e);
                }
            }
        );

        // write_file
        registerTool("write_file",
            "Write content to a file on disk.",
            schema("object",
                Map.of("path", Map.of("type", "string"), "content", Map.of("type", "string")),
                List.of("path", "content")),
            input -> {
                String path    = input.path("path").asText();
                String content = input.path("content").asText();
                try {
                    Files.writeString(Path.of(path), content);
                    return "Written " + content.length() + " bytes to " + path;
                } catch (IOException e) {
                    throw new RuntimeException("write_file failed: " + e.getMessage(), e);
                }
            }
        );

        // finish
        registerTool("finish",
            "Call when you have the final answer. Ends the loop.",
            schema("object", Map.of("answer", Map.of("type", "string")), List.of("answer")),
            input -> input.path("answer").asText()
        );
    }

    // ── Schema helper ─────────────────────────────────────────────────────────
    static JsonNode schema(String type, Map<String, Object> props, List<String> required) {
        try {
            ObjectNode n = JSON.createObjectNode();
            n.put("type", type);
            n.set("properties", JSON.valueToTree(props));
            ArrayNode req = JSON.createArrayNode();
            required.forEach(req::add);
            n.set("required", req);
            return n;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    // ── API call via ModelRouter (Claude / DeepSeek / OpenAI) ─────────────────
    static ModelRouter _router = null;

    static ModelRouter getRouter() {
        if (_router == null) _router = new ModelRouter();
        return _router;
    }

    /**
     * Calls the model via ModelRouter and converts the UnifiedResponse to a JsonNode
     * in Anthropic messages-API format so the existing loop logic works unchanged
     * regardless of which provider (Claude / DeepSeek / OpenAI) is active.
     */
    static JsonNode callApi(List<Map<String, Object>> messages, String system) throws Exception {
        ModelRouter.UnifiedResponse resp = getRouter().create(messages, buildToolSchemas(), system, 4096);

        // Build a JsonNode in the same shape the loop expects:
        // { "stop_reason": "...", "usage": { "input_tokens": N, "output_tokens": N }, "content": [...] }
        ObjectNode root    = JSON.createObjectNode();
        ObjectNode usage   = JSON.createObjectNode();
        ArrayNode  content = JSON.createArrayNode();

        root.put("stop_reason", resp.stopReason);
        usage.put("input_tokens",  resp.inputTokens);
        usage.put("output_tokens", resp.outputTokens);
        root.set("usage", usage);

        for (ModelRouter.ContentBlock block : resp.content) {
            ObjectNode b = JSON.createObjectNode();
            if (block.type == ModelRouter.BlockType.TEXT) {
                b.put("type", "text");
                b.put("text", block.text != null ? block.text : "");
            } else {
                b.put("type", "tool_use");
                b.put("id",   block.id   != null ? block.id   : "");
                b.put("name", block.name != null ? block.name : "");
                if (block.input != null) {
                    b.set("input", JSON.valueToTree(block.input));
                } else {
                    b.set("input", JSON.createObjectNode());
                }
            }
            content.add(b);
        }
        root.set("content", content);
        return root;
    }

    @SuppressWarnings("unchecked")
    static List<Map<String, Object>> buildToolSchemas() {
        List<Map<String, Object>> schemas = new ArrayList<>();
        for (ToolDef t : TOOLS.values()) {
            schemas.add(Map.of(
                "name",         t.name(),
                "description",  t.description(),
                "input_schema", t.schema()
            ));
        }
        return schemas;
    }

    // ── Core ReAct engine ─────────────────────────────────────────────────────
    static class ReactResult {
        String answer;
        Trace  trace;
        ReactResult(String answer, Trace trace) {
            this.answer = answer;
            this.trace  = trace;
        }
    }

    static ReactResult runReact(
        String task,
        String systemExtra,
        int maxIterations,
        Consumer<TraceStep> onStep
    ) throws Exception {

        String system = systemExtra == null || systemExtra.isBlank()
            ? REACT_SYSTEM
            : REACT_SYSTEM + "\n\n" + systemExtra;

        ArrayNode messages = JSON.createArrayNode();
        ObjectNode firstMsg = JSON.createObjectNode();
        firstMsg.put("role", "user");
        firstMsg.put("content", task);
        messages.add(firstMsg);

        Trace trace = new Trace();
        long startMs = System.currentTimeMillis();

        System.err.printf("[react] START: %s%n", truncate(task, 120));

        for (int i = 0; i < maxIterations; i++) {
            trace.iterations = i + 1;
            System.err.printf("[react] iteration %d%n", i + 1);

            JsonNode response = callApi(messages, system);

            long inputTokens  = response.path("usage").path("input_tokens").asLong();
            long outputTokens = response.path("usage").path("output_tokens").asLong();
            trace.totalTokens += inputTokens + outputTokens;

            JsonNode content    = response.path("content");
            String   stopReason = response.path("stop_reason").asText();

            // Append assistant turn
            ObjectNode assistantMsg = JSON.createObjectNode();
            assistantMsg.put("role", "assistant");
            assistantMsg.set("content", content);
            messages.add(assistantMsg);

            // end_turn — model responded with text
            if ("end_turn".equals(stopReason)) {
                for (JsonNode block : content) {
                    if ("text".equals(block.path("type").asText())) {
                        String answer = block.path("text").asText().trim();
                        if (!answer.isEmpty()) {
                            trace.steps.add(TraceStep.answer(answer));
                            trace.elapsedMs = System.currentTimeMillis() - startMs;
                            return new ReactResult(answer, trace);
                        }
                    }
                }
            }

            // tool_use
            if ("tool_use".equals(stopReason)) {
                ArrayNode toolResults = JSON.createArrayNode();

                for (JsonNode block : content) {
                    if (!"tool_use".equals(block.path("type").asText())) continue;

                    String   name  = block.path("name").asText();
                    String   id    = block.path("id").asText();
                    JsonNode input = block.path("input");

                    // Think
                    if ("think".equals(name)) {
                        String reasoning = input.path("reasoning").asText();
                        TraceStep step = TraceStep.thought(reasoning);
                        trace.steps.add(step);
                        if (onStep != null) onStep.accept(step);
                        System.err.printf("[react] 💭 %s%n", truncate(reasoning, 200));
                        ObjectNode tr = JSON.createObjectNode();
                        tr.put("type", "tool_result");
                        tr.put("tool_use_id", id);
                        tr.put("content", "OK");
                        toolResults.add(tr);
                        continue;
                    }

                    // Finish
                    if ("finish".equals(name)) {
                        String answer = input.path("answer").asText();
                        trace.steps.add(TraceStep.answer(answer));
                        trace.elapsedMs = System.currentTimeMillis() - startMs;
                        System.err.printf("[react] DONE — %d iters, %d tokens, %dms%n",
                            trace.iterations, trace.totalTokens, trace.elapsedMs);
                        return new ReactResult(answer, trace);
                    }

                    // Regular action
                    TraceStep actionStep = TraceStep.action(name, input);
                    trace.steps.add(actionStep);
                    if (onStep != null) onStep.accept(actionStep);
                    System.err.printf("[react] ⚡ %s(%s)%n", name, truncate(input.toString(), 200));

                    String  result  = null;
                    boolean isError = false;
                    ToolDef toolDef = TOOLS.get(name);

                    if (toolDef == null) {
                        result  = "Tool '" + name + "' not registered";
                        isError = true;
                    } else {
                        try {
                            result = toolDef.fn().apply(input);
                        } catch (Exception e) {
                            result  = "Tool error: " + e.getMessage();
                            isError = true;
                            System.err.printf("[react] ❌ %s: %s%n", name, result);
                        }
                    }

                    TraceStep obs = TraceStep.observation(result, isError);
                    trace.steps.add(obs);
                    if (onStep != null) onStep.accept(obs);
                    System.err.printf("[react] 👁 %s%n", truncate(result, 200));

                    ObjectNode tr = JSON.createObjectNode();
                    tr.put("type", "tool_result");
                    tr.put("tool_use_id", id);
                    tr.put("content", result);
                    tr.put("is_error", isError);
                    toolResults.add(tr);
                }

                ObjectNode userMsg = JSON.createObjectNode();
                userMsg.put("role", "user");
                userMsg.set("content", toolResults);
                messages.add(userMsg);
                continue;
            }

            System.err.printf("[react] unexpected stop_reason: %s%n", stopReason);
            break;
        }

        trace.elapsedMs = System.currentTimeMillis() - startMs;
        throw new RuntimeException(String.format(
            "ReAct loop did not finish within %d iterations (tokens: %d)",
            maxIterations, trace.totalTokens));
    }

    // ── Helpers ───────────────────────────────────────────────────────────────
    static String truncate(String s, int n) {
        return s.length() <= n ? s : s.substring(0, n) + "...";
    }

    // ── CLI entry point ───────────────────────────────────────────────────────
    public static void main(String[] args) throws Exception {
        String task = args.length > 0
            ? String.join(" ", args)
            : "Calculate the 47th Fibonacci number using bash, then explain the result.";

        ReactResult result = runReact(task, null, 30, step -> {
            switch (step.type()) {
                case "thought"     -> System.out.printf("%n💭 THOUGHT: %s%n",
                                            truncate(step.content(), 300));
                case "action"      -> System.out.printf("%n⚡ ACTION: %s(%s)%n",
                                            step.tool(),
                                            truncate(step.input() != null ? step.input().toString() : "", 200));
                case "observation" -> System.out.printf("%n%s OBSERVE: %s%n",
                                            step.error() ? "❌" : "👁",
                                            truncate(step.content(), 300));
            }
        });

        System.out.println("\n" + "=".repeat(60));
        System.out.println("FINAL ANSWER:");
        System.out.println(result.answer);
        System.out.println("=".repeat(60));
        System.out.printf("Iterations: %d | Tokens: %d | Time: %dms%n",
            result.trace.iterations, result.trace.totalTokens, result.trace.elapsedMs);

        System.out.println("\nFull trace:");
        ArrayNode traceJson = JSON.createArrayNode();
        result.trace.steps.forEach(s -> traceJson.add(s.toJson()));
        System.out.println(JSON.writerWithDefaultPrettyPrinter().writeValueAsString(traceJson));
    }
}

File: python/src/react_loop.py
"""
Production ReAct Loop — Python
Reason → Act → Observe, repeat until answer found.

Multi-provider via model_router: ANTHROPIC_API_KEY (Claude), DEEPSEEK_API_KEY,
or OPENAI_API_KEY depending on MODEL_PROVIDER env var.

pip install anthropic==0.50.0 openai==2.30.0 tavily-python==0.7.12 tenacity==9.1.2 structlog==24.4.0

export MODEL_PROVIDER=anthropic   # or deepseek | deepseek-openai | openai
export ANTHROPIC_API_KEY=sk-ant-...
export TAVILY_API_KEY=tvly-...
"""

import json
import os
import re
import subprocess
import sys
import time
from dataclasses import dataclass, field
from typing import Any, Callable

import structlog
from tenacity import retry, retry_if_exception_type, stop_after_attempt, wait_exponential

from model_router import ModelRouter, TextBlock, ToolUseBlock, make_tool_result_message

# ── Logging ───────────────────────────────────────────────────────────────────
structlog.configure(processors=[
    structlog.processors.TimeStamper(fmt="iso"),
    structlog.processors.add_log_level,
    structlog.dev.ConsoleRenderer(),
])
log = structlog.get_logger()


# ── ReAct trace types ─────────────────────────────────────────────────────────
@dataclass
class Thought:
    content: str
    type: str = field(default="thought", init=False)


@dataclass
class Action:
    tool: str
    input: dict
    type: str = field(default="action", init=False)


@dataclass
class Observation:
    content: str
    is_error: bool = False
    type: str = field(default="observation", init=False)


@dataclass
class Answer:
    content: str
    type: str = field(default="answer", init=False)


@dataclass
class Trace:
    steps: list = field(default_factory=list)
    total_tokens: int = 0
    iterations: int = 0
    elapsed_s: float = 0.0

    def add(self, step):
        self.steps.append(step)

    def to_dict(self):
        out = []
        for s in self.steps:
            if isinstance(s, Thought):
                out.append({"type": "thought", "content": s.content})
            elif isinstance(s, Action):
                out.append({"type": "action", "tool": s.tool, "input": s.input})
            elif isinstance(s, Observation):
                out.append({"type": "observation", "content": s.content, "error": s.is_error})
            elif isinstance(s, Answer):
                out.append({"type": "answer", "content": s.content})
        return out


# ── Tool registry ─────────────────────────────────────────────────────────────
TOOL_REGISTRY: dict[str, Callable] = {}
TOOL_SCHEMAS: list[dict] = []


def tool(name: str, description: str, schema: dict):
    """Decorator to register a tool."""
    def decorator(fn: Callable):
        TOOL_REGISTRY[name] = fn
        TOOL_SCHEMAS.append({
            "name": name,
            "description": description,
            "input_schema": schema,
        })
        return fn
    return decorator


# ── Built-in tools ──────────────────────────────────────────────────────────────

@tool(
    name="think",
    description="Use this to reason step by step before acting. Does not call any external system.",
    schema={
        "type": "object",
        "properties": {"reasoning": {"type": "string", "description": "Your chain of thought"}},
        "required": ["reasoning"],
    },
)
def think(reasoning: str) -> str:
    return "OK"


# Lazily-constructed Tavily client (avoids hard failure if TAVILY_API_KEY unset at import time)
_tavily_client = None


def _get_tavily():
    global _tavily_client
    if _tavily_client is None:
        from tavily import TavilyClient
        api_key = os.environ.get("TAVILY_API_KEY")
        if not api_key:
            raise RuntimeError(
                "TAVILY_API_KEY is not set. Get a free key at https://app.tavily.com "
                "and `export TAVILY_API_KEY=tvly-...`"
            )
        _tavily_client = TavilyClient(api_key=api_key)
    return _tavily_client


@tool(
    name="search",
    description="Search the live web for current information. Returns titles, URLs, and content snippets.",
    schema={
        "type": "object",
        "properties": {
            "query": {"type": "string", "description": "Search query"},
            "num_results": {"type": "integer", "description": "Number of results (default 5, max 10)"},
            "search_depth": {"type": "string", "enum": ["basic", "advanced"], "description": "basic (fast) or advanced (deeper, slower)"},
        },
        "required": ["query"],
    },
)
def search(query: str, num_results: int = 5, search_depth: str = "basic") -> str:
    client = _get_tavily()
    response = client.search(
        query=query,
        max_results=min(max(num_results, 1), 10),
        search_depth=search_depth,
        include_answer=True,
    )

    parts = []
    if response.get("answer"):
        parts.append(f"Direct answer: {response['answer']}")

    for i, r in enumerate(response.get("results", []), 1):
        content = r.get("content", "")[:600]
        parts.append(f"[{i}] {r.get('title', '')}\nURL: {r.get('url', '')}\n{content}")

    return "\n\n".join(parts) if parts else "No results found."


@tool(
    name="fetch_url",
    description="Fetch and extract the readable text content of a specific URL.",
    schema={
        "type": "object",
        "properties": {"url": {"type": "string"}},
        "required": ["url"],
    },
)
def fetch_url(url: str) -> str:
    try:
        # Prefer Tavily's extract endpoint for clean, agent-ready content
        client = _get_tavily()
        response = client.extract(urls=[url])
        results = response.get("results", [])
        if results:
            return results[0].get("raw_content", "")[:8000]
    except Exception:
        pass

    # Fallback: raw HTTP fetch + HTML strip
    import urllib.request
    with urllib.request.urlopen(url, timeout=10) as r:
        raw = r.read().decode("utf-8", errors="replace")
    clean = re.sub(r"<script.*?</script>", " ", raw, flags=re.S | re.I)
    clean = re.sub(r"<style.*?</style>", " ", clean, flags=re.S | re.I)
    clean = re.sub(r"<[^>]+>", " ", clean)
    clean = re.sub(r"\s+", " ", clean).strip()
    return clean[:8000]


@tool(
    name="run_python",
    description="Execute Python code in a sandboxed subprocess. Returns combined stdout + stderr. Timeout 15s.",
    schema={
        "type": "object",
        "properties": {"code": {"type": "string"}},
        "required": ["code"],
    },
)
def run_python(code: str) -> str:
    result = subprocess.run(
        [sys.executable, "-c", code],
        capture_output=True, text=True, timeout=15,
    )
    out = result.stdout or ""
    err = result.stderr or ""
    combined = (out + err).strip()
    return combined[:8000] if combined else "(no output)"


@tool(
    name="read_file",
    description="Read a file from disk (UTF-8 text).",
    schema={
        "type": "object",
        "properties": {"path": {"type": "string"}},
        "required": ["path"],
    },
)
def read_file(path: str) -> str:
    with open(path, "r", encoding="utf-8") as f:
        return f.read()[:16000]


@tool(
    name="write_file",
    description="Write content to a file on disk, creating parent directories if needed.",
    schema={
        "type": "object",
        "properties": {
            "path": {"type": "string"},
            "content": {"type": "string"},
        },
        "required": ["path", "content"],
    },
)
def write_file(path: str, content: str) -> str:
    import pathlib
    p = pathlib.Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    with open(p, "w", encoding="utf-8") as f:
        f.write(content)
    return f"Written {len(content)} bytes to {path}"


@tool(
    name="finish",
    description="Call this when you have the final answer. Ends the loop.",
    schema={
        "type": "object",
        "properties": {"answer": {"type": "string"}},
        "required": ["answer"],
    },
)
def finish(answer: str) -> str:
    return answer


# ── Model router (multi-provider: Claude / DeepSeek / OpenAI) ────────────────
_router = ModelRouter()


@retry(
    retry=retry_if_exception_type(Exception),
    wait=wait_exponential(multiplier=1, min=2, max=60),
    stop=stop_after_attempt(6),
)
def _call_model(messages: list, system: str):
    return _router.create(messages=messages, tools=TOOL_SCHEMAS, system=system, max_tokens=4096)


# ── ReAct system prompt ───────────────────────────────────────────────────────
REACT_SYSTEM = """You are an autonomous agent operating in a ReAct loop (Reason + Act + Observe).

For every task:
1. Use the `think` tool to reason about what to do next before acting.
2. Call the appropriate tool to act.
3. Observe the result and reason again.
4. Repeat until you have a complete, verified answer.
5. Call `finish` with your final answer when done.

Rules:
- Always think before acting. Never skip the think step.
- If a tool errors, reason about why and try a different approach.
- Do not guess. If unsure, use `search` or `fetch_url`.
- Be thorough. Do not call finish until the task is fully complete.
"""


# ── Core ReAct engine ─────────────────────────────────────────────────────────
def run_react(
    task: str,
    system_extra: str = "",
    max_iterations: int = 30,
    on_step: Callable[[Any], None] | None = None,
) -> tuple[str, Trace]:
    """
    Run a full ReAct loop.

    Args:
        task:           The task/question to solve.
        system_extra:   Additional system instructions appended to base prompt.
        max_iterations: Hard iteration cap.
        on_step:        Optional callback called after each Thought/Action/Observation.

    Returns:
        (final_answer: str, trace: Trace)
    """
    system = REACT_SYSTEM + ("\n\n" + system_extra if system_extra else "")
    messages: list[dict] = [{"role": "user", "content": task}]
    trace = Trace()
    start = time.time()

    log.info("react.start", task=task[:120], provider=_router.provider, model=_router.model)

    for i in range(max_iterations):
        trace.iterations = i + 1
        log.info("react.iteration", n=i + 1)

        response = _call_model(messages, system)
        trace.total_tokens += response.usage.input_tokens + response.usage.output_tokens
        messages.append(response.to_assistant_message())

        if response.stop_reason == "end_turn":
            for block in response.content:
                if isinstance(block, TextBlock) and block.text.strip():
                    answer = block.text.strip()
                    trace.add(Answer(answer))
                    trace.elapsed_s = round(time.time() - start, 2)
                    log.info("react.done", via="end_turn", **_trace_stats(trace))
                    return answer, trace

        if response.stop_reason == "tool_use":
            for block in response.content:
                if not isinstance(block, ToolUseBlock):
                    continue

                tool_name = block.name
                tool_input = block.input

                if tool_name == "think":
                    thought = Thought(tool_input.get("reasoning", ""))
                    trace.add(thought)
                    if on_step:
                        on_step(thought)
                    log.info("react.thought", content=thought.content[:200])
                    messages.append(make_tool_result_message(block.id, "OK"))
                    continue

                if tool_name == "finish":
                    answer = tool_input.get("answer", "")
                    trace.add(Answer(answer))
                    trace.elapsed_s = round(time.time() - start, 2)
                    log.info("react.done", via="finish", **_trace_stats(trace))
                    return answer, trace

                action = Action(tool=tool_name, input=tool_input)
                trace.add(action)
                if on_step:
                    on_step(action)
                log.info("react.action", tool=tool_name, input=str(tool_input)[:200])

                is_error = False
                try:
                    if tool_name not in TOOL_REGISTRY:
                        raise ValueError(f"Tool '{tool_name}' not registered")
                    result = str(TOOL_REGISTRY[tool_name](**tool_input))
                except Exception as e:
                    result = f"Tool error ({type(e).__name__}): {e}"
                    is_error = True
                    log.error("react.tool_error", tool=tool_name, error=str(e))

                observation = Observation(content=result, is_error=is_error)
                trace.add(observation)
                if on_step:
                    on_step(observation)
                log.info("react.observation", tool=tool_name,
                         result_len=len(result), error=is_error)

                messages.append(make_tool_result_message(block.id, result, is_error=is_error))

            continue

        log.warning("react.unexpected_stop", stop_reason=response.stop_reason)
        break

    trace.elapsed_s = round(time.time() - start, 2)
    raise RuntimeError(
        f"ReAct loop did not finish within {max_iterations} iterations. "
        f"Tokens used: {trace.total_tokens}"
    )


def _trace_stats(trace: Trace) -> dict:
    return {
        "iterations": trace.iterations,
        "tokens": trace.total_tokens,
        "elapsed_s": trace.elapsed_s,
        "steps": len(trace.steps),
    }


# ── CLI entry point ───────────────────────────────────────────────────────────
if __name__ == "__main__":
    task = " ".join(sys.argv[1:]) if len(sys.argv) > 1 else (
        "Calculate the 47th Fibonacci number using code, then explain the result."
    )

    def print_step(step):
        if isinstance(step, Thought):
            print(f"\n💭 THOUGHT: {step.content[:300]}")
        elif isinstance(step, Action):
            print(f"\n⚡ ACTION: {step.tool}({json.dumps(step.input)[:200]})")
        elif isinstance(step, Observation):
            marker = "❌" if step.is_error else "👁"
            print(f"\n{marker} OBSERVE: {step.content[:300]}")

    answer, trace = run_react(task, on_step=print_step)

    print("\n" + "=" * 60)
    print("FINAL ANSWER:")
    print(answer)
    print("=" * 60)
    print(f"Iterations: {trace.iterations} | Tokens: {trace.total_tokens} | Time: {trace.elapsed_s}s")
    print("\nFull trace (JSON):")
    print(json.dumps(trace.to_dict(), indent=2))
